EIPS/EIPs/eip-5075.md

```diff
@@ -0,0 +1,236 @@
+  
+Contracts SHALL replace all ETH and Token based transfers with transferL where the transfers are to unknown external addresses or are publicly accessible.
+  
+```function transferL(address _to, uint256 _amount, address _token) public returns (bool success)```
+
+  
+**getLimitLeft**
+
+
+Returns the amount allowed within the rate limited transfer allowance for the current time with current outflow.
+
+```function getLimitLeft(address token) public view returns (uint256 rateLeft)```
+  
+  
+**whitelist**
+
+
+Allows an address or addresses to whitelist other addresses as msg.sender to not be rate limited.
+OPTIONAL - This allows for an owner or other mechanism to allow contracts such as YEARN to be able to move positions or manage positions without affecting or being affected by the rate limit and out flows. Use with caution and for contracts with known deployed bytecode and that have been verified to not be malicious or able to be maliciously used to allow for getting around the rate limit during hacks. Note remember to restrict the address able to use the whitelist and removeWhitelist fn.
+  
+```function whitelist(address addrs) public```
+
+```function removeWhitelist(address addrs) public```
+## Rationale
+By limiting token outflows at a per token level, malicious interactions can be limited in their damage as all net outflows are tracked and limited for all users based on a realtime allowance based on acceptable outflow relative to total amount held and within a given timeframe. The decision to track at the token level was for simplicity, gas efficiency and not needing to itself have vulnerabilities with oracles and accounting to track as a pool vs per asset. Whitelisting allows for contracts like yearn to be compatible if the destination is  a known address not likely to be used in an attack. Dynanmic rate limits will help not stop but limit the possible damage, as all outflows need to use a transfer fn to steal assets, which this stops if reaches the rate limit.
+
+## Backwards Compatibility
+This will be non backwards compatible with contracts that are NOT whitelisted which assume they can just withdraw large positions, if those positions are large relative to the target contracts net position by all users. Should this be the case the contract can be fully compatible by simply being whitelisted, skipping all rate limit checks.
+
+## Reference Implementation
+
+```
+  // SPDX-License-Identifier: CC0
+pragma solidity ^0.8;
+
+import"./IERC20.sol";
+
+contract rateLimit {
+    mapping(address => TokenInfo) public infoL;
+    struct TokenInfo {
+        uint32 timestamp; //tracks most recent send
+        uint224 rate; //tracks amount sent for the last timeLimit time
+    }
+
+    uint256 public timeLimit = 3600; //how long in seconds to limit within, recommend 1h = 3600
+    uint16 public rateLimit = 1000; //the basis points (00.0x%) to allow as the max sent within the last timeLimit time
```

Comment:
nit: `rate` is in amount unit, but `rateLimit` is in bps which is a bit confusing to read. Maybe rename `rate` to `outflow` (because an amount is not a rate?)

Reply:
Noted, will do